packages/rich-text/src/helpers/normalizers.ts

```diff
@@ -0,0 +1,37 @@
+import { NodeEntry, Element } from 'slate';
+import { PlateEditor } from '@udecode/plate-core';
+
+import { CustomElement } from '../types';
+
+export type Normalizer = (
+  editor: PlateEditor,
+  entry: NodeEntry<CustomElement>
+) => boolean | undefined;
+
+/**
+ * Injects a custom element normalization handler.
+ *
+ * Handlers must explicity return "true" to indicate a pass.
+ * A pass is when a handler didn't encounter any normalization
+ * paths (i.e. all nodes were valid)
+ *
+ * Important read:
+ * https://docs.slatejs.org/concepts/11-normalizing#multi-pass-normalizing
```

Comment:
I don't get it. How exactly will be gain some performance boost?

The `return true` in this case is not for performance benefit but this:

> You might at first think this is odd, because with the return there, the original normalizeNodes will never be called, and the built-in constraints won't get a chance to run their own normalization.
>
> But, there's a slight "trick" to normalizing.
> 
> When you do call Editor.unwrapNodes, you're actually changing the content of the node that is currently being normalized. So even though you're ending the current normalization pass, by making a change to the node you're kicking off a new normalization pass. This results in a sort of recursive normalizing.
This multi-pass characteristic makes it much easier to write normalizations, because you only ever have to worry about fixing a single issue at once, and not fixing every possible issue that could be putting a node in an invalid state.

Reply:
Btw, we don't "control" when a normalization pass kicks in. Even if we return `true` and continue recursively calling `normalizeNode`, slate will still kick in a new normalization because the changes we did

Reply:
As I understand, `return true` results in another full normalization cycle (that's why it's called multi-pass).

If you could prevent this and just finish the current cycle (or "pass") instead, you'd reduce the total number of node checks.
But I guess even if you do this there would be another pass anyways, as you touched a node? No idea, this goes way too deep into Slate internals.

Reply:
> But I guess even if you do this there would be another pass anyways, as you touched a node?

Yes. Exactly. Slate controls when kick a new pass (when a node is touched) 

We control when to finish the current pass. So it could be better to prevent bugs & performance-wise to exit early. Does that make sense now?